rfc/20250728-opentofu-providers/03-provider-sdk.md

```diff
@@ -0,0 +1,618 @@
+
+**3. Language-Idiomatic Design**
+Each SDK follows the conventions and patterns of its target language ecosystem, making provider development feel natural to developers already familiar with that language.
+
+**4. Progressive Disclosure**
+Simple use cases require minimal code, while complex scenarios remain possible through advanced SDK features and escape hatches.
+
+**5. Automatic Features**
+The SDK automatically should provide managed resources, data sources, documentation generation, capability negotiation, and protocol handling, etc. without requiring developer intervention.
+
+## SDK Architecture
+
+### Transport Abstraction
+
+All SDKs use a transport abstraction that handles the underlying protocol communication. The initial implementation focuses on stdio transport, with future support for other transports:
+
+```typescript
+// TypeScript example
+import { Provider, StdioTransport } from '@opentofu/provider-sdk';
+
+const provider = new Provider({ name: "custom", version: "1.0.0" });
+new StdioTransport().connect(provider);
+```
+
+### Protocol Translation Layer
+
+The SDK acts as a translation layer between the developer's simple methods and something that could fulfil the `providers.interface` specification. See [02-provider-client-library.md](./02-provider-client-library.md) for more details.
+
+For example a developer should be able to write the following:
+```typescript
+provider.resource("s3_bucket", {
+  ...
+  methods: {
+    async create(config) {
+      const result = await my_s3_client.createBucket(config);
+      return { id: result.id, state: { ...config, id: result.id } };
+    }
+  }
+}
+```
+
+And the SDK translates the calls to/from such functions:
+- `GetProviderSchema` responses with proper schema definitions
+- `PlanResourceChange` handling with unknown value management
+- `ApplyResourceChange` execution with error handling
+- State persistence and retrieval logic
+
+## Multi-Language Implementation
+
+### TypeScript SDK
+
+> [!NOTE]
+> The proposal for typescript may seem more comlpete as it is what I am most familiar with. Other language examples are shown to define the pattern of implementation as an example, and not as something set in stone.
+
+The TypeScript SDK could leverage Zod for schema validation and provides a simple experience of defining items on the provider. For example:
+
+```typescript
+import { z } from 'zod';
+import { Provider, StdioTransport } from '@opentofu/provider-sdk';
+
+const provider = new Provider({
+  name: "my-custom-aws-s3",
+  version: "0.1.0",
+});
+
+// Schema-first resource definition
+const s3BucketSchema = z.object({
+  bucket: z.string(),
+  region: z.string().default("us-east-1"),
+  versioning: z.boolean().default(false),
+  tags: z.record(z.string()).optional(),
+  // Computed fields
+  arn: z.string().optional(),
+  id: z.string().optional(),
+});
+
+provider.resource("s3_bucket", {
```

Comment:
How would you imagine this handling "plan"?  What if tags are a dict of unknown or map[string]unknown.  Would you be able to validate the keys?

This is probably one of those "escape hatch" scenarios?

Reply:
The Terraform plugin framework handles this with a concept called [plan modifiers](https://developer.hashicorp.com/terraform/plugin/framework/resources/plan-modification), and I think its design is a nice compromise observing that there are several situations that arise often for lots of providers, like "any change to these attributes requires replace instead of update", but there are also sometimes custom rules that are best written inline.

The specific API there has a lot of type-related complexity because Go is statically-typed, but I expect that a dynamically-typed version of that could be a lot lighter.